src/cargo/util/graph.rs

```diff
@@ -1,52 +1,219 @@
-    pub fn link(&mut self, node: N, child: N) -> &mut E {
-        self.nodes
-            .entry(node)
-            .or_insert_with(im_rc::OrdMap::new)
-            .entry(child)
-            .or_insert_with(Default::default)
+    /// connect `node`to `child` with out associating any data.
+    /// Note that if this and `add_edge` are used on the same graph
+    ///      odd things may happen when `reset_to` is called.
+    pub fn link(&mut self, node: N, child: N) {
+        use indexmap::map::Entry;
+        // IndexMap happens to do exactly what we need to keep the ordering correct.
+        match self
+            .nodes
+            .entry(node.clone())
+            .or_insert_with(IndexMap::new)
+            .entry(child.clone())
+        {
+            Entry::Vacant(entry) => {
+                // add the new edge, and link and fix the new link count
+                self.edges.push(EdgeLink {
+                    value: None,
+                    next: None,
+                    previous: None,
+                });
+                let edge_index: EdgeIndex = self.edges.len() - 1;
+                entry.insert(edge_index);
+            }
+            Entry::Occupied(_) => {
+                // this pare is already linked
+            }
+        };
+    }
+
+    /// connect `node`to `child` associating it with `edge`.
+    /// Note that if this and `link` are used on the same graph
+    ///      odd things may happen when `reset_to` is called.
+    pub fn add_edge(&mut self, node: N, child: N, edge: E) {
+        use indexmap::map::Entry;
+        let edge = Some(edge);
+        // IndexMap happens to do exactly what we need to keep the ordering correct.
+        match self
+            .nodes
+            .entry(node.clone())
+            .or_insert_with(IndexMap::new)
+            .entry(child.clone())
+        {
+            Entry::Vacant(entry) => {
+                // add the new edge, and link and fix the new link count
+                self.edges.push(EdgeLink {
+                    value: edge,
+                    next: None,
+                    previous: None,
+                });
+                let edge_index: EdgeIndex = self.edges.len() - 1;
+                entry.insert(edge_index);
+            }
+            Entry::Occupied(entry) => {
+                // this pare is already linked
+                let mut edge_index = *entry.get();
+                loop {
+                    // follow the linked list
+                    if self.edges[edge_index].value == edge {
+                        return;
+                    }
+                    if self.edges[edge_index].next.is_none() {
+                        // we found the end, add the new edge
+                        self.edges.push(EdgeLink {
+                            value: edge,
+                            next: None,
+                            previous: Some(edge_index),
+                        });
+                        let new_index: NonZeroEdgeIndex =
```

Comment:
Is there much gained here from using `NonZero*` instead of just `usize`? I find the `NonZero*` stuff here to be somewhat wordy and detracting from how it's just a pointer into the array

Reply:
8 bytes per edge. Not worth the boilerplate. Removed.